Connect login/signup/reset functionalities to backend

19ec94 opened the pull request:
This pull request will,
 - create a reusable custom react FormElement to be used in Login.jsx, Signup.jsx, and Forgot.jsx files,
 - do the following modifications in Login.jsx file:
   - remove existing state capturing mechanism (`action/setAction`),
   - de-couple login from signup (and simplify the codebase),
   - create a `form` element and a submit `button`,
   - implement submission logic using `axios` library,
   - add navigation links to the signup and the reset pages,
 - do the following modifications in Signup.jsx file:
   - remove existing state capturing mechanism (`action/setAction`),
   - create a `form` element and submit `button`,
   - implement submission logic using `axios` library,
   - add a navigation link to the signup page,
 - do the following modifications in Forgot.jsx:
   - remove existing state capture mechanics (`action/setAction`),
   - create a boolean-type `isSubmitted` variable to keep track of state of the form,
   - add a navigation link to login page and reset page,
 - do the following modifications in App.js:
   - create dedicated routes for signup, signup, reset,
   - change `import` statements for enhanced control,
 - remove unnecessary style elements from the Login.css file.

19ec94 committed:
Edit logIn to Login

19ec94 committed:
Update package - install axios to make http request

19ec94 committed:
Refactor Form element in the login/signup page

In order to simplify the codebase, the following actions are proposed.
- Add a form element in login/signup page. This makes it easy to submit
  login/sigup/reset data to the backend.
- Create a React element called FormElement and store it in the file
  FormElement.jsx.
- Import the contents in FormElement.jsx in Login.jsx component and use
  it to create input fields in the form element.

19ec94 committed:
Change import statements

In the interest of enhancing flexibility and security, the following
changes are proposed:
- Explicitly import only the necessary files which contain required react
  components

19ec94 committed:
Seperate Login and Signup functionality

To facilitate separate routing mechanisms for login and signup in
backend development and to reduce the complexity in the frontend
codebase the following changes are proposed:
  - Create seperate forms for login and signup
  - Add navigation links between login and signup. This makes is easy to
    go back-and-forth between these two seperate forms.

19ec94 committed:
Remove "action" variable

Since the functionalities of signup and login are seperated from one
another, "action" variable is not needed to capture the current state of
the form.

19ec94 committed:
Replace div container submit by a dedicated submit button element

To submit the signup form data separate function, the following changes
are proposed:
  - replace the generic div container with onClick callback function by
    a more specific html element - a button type of submit.
Upon form submission, to handle the submission logic
  - add a separate handle function.

Minor asthetic change:
  - Remove passwords container div element and its styling element

19ec94 committed:
Remove the password styling element which is not needed

19ec94 committed:
Replace the generic div by a button of type submit

Following the changes proposed in
commit 5578d635ac6f6fdfc5316b679ec2b7978ab0d4f8 ,
 - make same changes in this file

19ec94 committed:
Add varibales to capture the current state

To capture the user input into a variable the following modifications
are proposed:
  - add `value` attribute to capture the value,
  - add `onChange` attribute to pass handleChange function which updates
    the values on change.

19ec94 committed:
Add capture mechanism to capture data to be sent to backend

In order to send data to backend, the following changes are proposed:
  - use `userState` to capture varibales
  - add `handleChange` to update variables on change (e.g, user input)
  - import `axios` library to make http request to backend
  - edit `handleChange` to trigger submission logic and process error
    response
  - Following the changes in commit
    e259dcce9dfb8632138d584821345c95eccb9dc6, make necessary
    modifications in FormElement.

19ec94 committed:
Add client-side form validation

To prevent empty form submission,
  - add basic check for empty string, null, undefined values.
To provide feedback to the user,
  - add a <p></p> element above submit button to display error message.

19ec94 committed:
Edit display type of submit container

Following the changes in commit 7a992b70c65ac1d5433afa1a70dc45d2e392e987
change style configuration

Minor change: remove unncessary code portion

19ec94 committed:
Update dependencies

19ec94 committed:
Add proxy connection

19ec94 committed:
Change API end points upon http request

Following the commit a129cc2049750fdc3645168b3f38456dc2f175bb,
 - Change API end point from specific url to more generic one

19ec94 committed:
Add dev dependency

Add the following dev devpendency to package.json file to get rid of the pesky warnings
  - "@babel/plugin-proposal-private-property-in-object": "^7.21.0"

19ec94 committed:
Resolve merge conflict

19ec94 committed:
Simplify reset password operation

To follow the same approach used in login and signup forms, the
following changes are proposed:
  - add the custom FormElement for email field
  - remove `action` variable to capture the state. Instead, create a
    boolean-type variable `isSubmitted` to keep track of the submission state.
  - in case of no email received, direct the user back to reset page by
    providing the same link to original reset page.

19ec94 committed:
Remove uncessary styles

19ec94 committed:
Add a nav link to login page

To link reset and login pages,
 - add a nav link to login page before submitting the reset form. The
   user may have rememebered their login details.
Edit texts to improve clarity.

19ec94 committed:
Add state variables to track error

For better error management, more robust error-checking must be done.
In this commit,
 - add basic error checking to prevent empty submissions

19ec94 committed:
Add state variable `error` to track error

In this commit,
 - add basic error checking to ensure no-empty submissions are allowed
For better error management, more robust error-checking must be done.

19ec94 committed:
Merge branch 'main' into connect-login-backend

19ec94 committed:
Merge branch 'main' into connect-login-backend

19ec94 committed:
Merge branch 'connect-login-backend' of github.com:19ec94/interactive-3d-web-design into connect-login-backend

19ec94 committed:
Add Context API to manage Auth related global variables

To make authentication related varibales available across serveral react
components we choose to use Context API to create and manage globals
variables such as isLoggedIn and functions login(), logout()

Choose sessionStorage to store tokens to avoid automatic logout upon
refreshing the page. Note that, sessionStorage is only valid for a
single tab.

19ec94 committed:
Import Conext API for AuthContext

This commit will
 - Remove Routes element from header element
 - Wrap header and Routes elements by AuthContext element to provide
   access to authentication-related global variables

19ec94 committed:
Update isLoggedIn status upon successful login/signup

19ec94 committed:
Render components based on login status

Display /, /About, /Login routes if the user is not logged in
Display all the pages if the user is logged in

19ec94 committed:
Move react scripts from dependencies to devDependencies

To avoid false warnings, move react-scripts from dependencies to
devDepenedencies as suggessted here
https://stackoverflow.com/questions/71282206/github-dependabot-alert-inefficient-regular-expression-complexity-in-nth-check

19ec94 committed:
Navigate to dashboard after sucessfull login/signup

19ec94 committed:
Create skeleton page for levels

19ec94 committed:
Make levels, dashboard protected routes

If the user access levels or dashboard endpoit and is not logged in,
then re-direct the user to login page.

19ec94 committed:
Add dashboard page and its style

19ec94 committed:
Direct the user to dashboard after successful login/signup

19ec94 committed:
Add path to dashboard

19ec94 committed:
Add scoreboard link to unprotected access group

19ec94 committed:
Edit sample text

19ec94 committed:
Rename file for consistency

19ec94 committed:
Add autocomplete attribute

19ec94 committed:
Replace levels by dashboard in navbar

Rationale: levels don't need to be accessed  directly from nav bar.
Levels should be accessed from dashboard using internal linking.

19ec94 committed:
Redirect logged in user to dashboard

Rationale:
Don't let logged in user to access login page again. Redirect the
already logged in user to dashboard.

19ec94 committed:
Reorgonise files

To avoid confusion between pages and components, we stick to components.
Conceptually, each page is just a big React component that comprises
several subcomponents.

19ec94 committed:
Re-order import statements alphabatically

19ec94 committed:
Add style marging for mobile view <480px

19ec94 committed:
Remove login link from navbar after log in

19ec94 committed:
Format file spacing

19ec94 committed:
Change  for mobile view

19ec94 committed:
Rename Forgot to Reset to reset login details

19ec94 committed:
Import and define paths alphabatically

19ec94 committed:
Add gloabl variable to hold session token

19ec94 committed:
Redirect logged in user to dashboard

Rationale:
 If the user is already logged in, visiting login/signup/reset endpoint
 is unnecessary. So, redirect the user if the user is already logged in.

19ec94 committed:
Add instructions to start and run the server

19ec94 committed:
Add a container and left margin to logo

19ec94 committed:
Install fontawesome webfonts

Source:
https://fontawesome.com/docs/web/use-with/react/

19ec94 committed:
Redirect undefined routes to /about page

19ec94 committed:
Replace image icons by fontawesome icons

Rationale:
 - Fontawesome provides an array of web fonts and icons and symbols.
 - Fontawesome Icons are easier to style just like regular html elements.

joserigel commented on src/components/Login.jsx:
```diff
@@ -0,0 +1,120 @@
+       * ": "error message:}}
+       */
+      console.log('error', error); // debugging
+      if (error.response.data.error.message){ // debugging
```
`response.data` gives a null pointer error, perhaps a simple solution would be `error?.response?.data?.error?.message`

joserigel replied:
`response.data` gives a null pointer error, perhaps `response?.data` ? but I see that you use something like this on other lines as well

joserigel commented on src/components/Navbar.css:
```diff
@@ -1,3 +1,6 @@
+* {
+  border: solid red 1px;
```
debugging css should not be included in the pull request :wink:

joserigel commented on src/components/Navbar.jsx:
```diff
@@ -35,24 +42,36 @@ export const Navbar = () => {
             console.log();
```
empty `console.log` ?

joserigel commented on src/components/Navbar.jsx:
```diff
@@ -1,20 +1,27 @@
+  // Define a variable to store login status. If logged in, display all
+  // navigation link, else display only about, login links.
+  const {isLoggedIn} = useAuth();
+  console.log('navbar element', isLoggedIn);//Debugging
```
refrain from putting debugging residues into the pull request, but since this is still early in the development process it's fine :laughing:

joserigel reviewed (changes requested):
The functionality works as I tested it, but there are still changes that might make it more polished to have it on the main branch

19ec94 committed:
Replace Create-React-App by Vite

Vite is strict about file name extensions. So, rename js files as jsx.

19ec94 committed:
Link application entry point as required by vite

19ec94 committed:
Create a react component for Profile in dashboard

19ec94 committed:
Simplify dashboard

Rationale:
 - Dashboard is going to content several sections in the future.
   Therefore, following react priciple, create seperate files for each
   comoponent.
This commit will
 - remove profile code from dashboard
 - import Profile react component

19ec94 committed:
Simplify global data and sessionStorage

Rationale:
 - collect data related to each session into one varibale - sessionData.

19ec94 committed:
Remove styles related to profile element

19ec94 committed:
Update the login logic

Introduce a condition to check if the sessionData is not empty upon
endpoint invokation. If empty, initialise the value. If not, assign the
value form sessionStorage.

19ec94 committed:
Remove debug log

19ec94 committed:
Remove debug log

19ec94 committed:
Update profile data

19ec94 committed:
Add box-sizing property to every html element

Rationale:
 - border-box property makes it easier to manage width and height of html
   element in comparison to default properties.

19ec94 committed:
Add game links

19ec94 committed:
Replace button text by fontawesome icon

19ec94 committed:
Add Styles to input field elements

- fix the width of label, input/span, button elements by presecribing
  width in terms of percentage of view port

19ec94 committed:
Fix style of profile element for large screen

joserigel reviewed (approved):
LGTM
